src/main/java/com/comphenix/protocol/injector/collection/OutboundPacketListenerSet.java

```diff
@@ -0,0 +1,88 @@
+package com.comphenix.protocol.injector.collection;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+
+import javax.annotation.Nullable;
+
+import com.comphenix.protocol.PacketType;
+import com.comphenix.protocol.ProtocolLibrary;
+import com.comphenix.protocol.concurrent.PacketTypeListenerSet;
+import com.comphenix.protocol.error.ErrorReporter;
+import com.comphenix.protocol.events.ListenerPriority;
+import com.comphenix.protocol.events.ListeningWhitelist;
+import com.comphenix.protocol.events.PacketContainer;
+import com.comphenix.protocol.events.PacketEvent;
+import com.comphenix.protocol.events.PacketListener;
+
+public class OutboundPacketListenerSet extends PacketListenerSet {
+
+	public OutboundPacketListenerSet(PacketTypeListenerSet mainThreadPacketTypes, ErrorReporter errorReporter) {
+		super(mainThreadPacketTypes, errorReporter);
+	}
+
+	@Override
+	protected ListeningWhitelist getListeningWhitelist(PacketListener packetListener) {
+		return packetListener.getSendingWhitelist();
+	}
+
+	@Override
+	public void invoke(PacketEvent event, @Nullable ListenerPriority priorityFilter) {
+		super.invoke(event, priorityFilter);
+
+		if (event.getPacketType() == PacketType.Play.Server.BUNDLE && !event.isCancelled()) {
+			// unpack the bundle and invoke for each packet in the bundle
+			Iterable<PacketContainer> packets = event.getPacket().getPacketBundles().read(0);
+			List<PacketContainer> outPackets = new ArrayList<>();
+			for (PacketContainer subPacket : packets) {
+				if (subPacket == null) {
+					ProtocolLibrary.getPlugin().getLogger().log(Level.WARNING,
+							"Failed to invoke packet event "
+									+ (priorityFilter == null ? "" : ("with priority " + priorityFilter))
+									+ " in bundle because bundle contains null packet: " + packets,
+							new Throwable());
+					continue;
+				}
+				PacketEvent subPacketEvent = PacketEvent.fromServer(this, subPacket, event.getNetworkMarker(),
+						event.getPlayer());
+				super.invoke(subPacketEvent, priorityFilter);
+
+				if (!subPacketEvent.isCancelled()) {
+					// if the packet event has been cancelled, the packet will be removed from the
+					// bundle
+					PacketContainer packet = subPacketEvent.getPacket();
+					if (packet == null) {
+						ProtocolLibrary.getPlugin().getLogger().log(Level.WARNING,
```

Comment:
here & below i think we'd want to call out which plugin is doing this so we can tell them not to

Reply:
As mentioned above, we need to throw an exception here because null packets are problematic. While creating a null packet isn't typically possible, the `readObject` call doesn't perform a null check, making it a rare but potential occurrence.

Reply:
generally it's a good assumption that someone is going to do something weird and introduce a null packet. better to be safe than sorry